File: android/pte-im-sdk/src/main/java/com/ptelive/im/PteIMCommerce.kt
package com.ptelive.im

import org.json.JSONArray
import org.json.JSONObject

/**
 * First-party IM Commerce extension. It shares the SDK's UserSig but owns no
 * WebSocket: room delivery is performed by Commerce -> IM scene outbox.
 */
class PteIMCommerce internal constructor(private val sdk: PteIMSDK) {
  fun capabilities(callback: (Result<PteIMCommerceCapability>) -> Unit) = request(callback) {
    val data = sdk.postCommerceJson("/v1/commerce/capabilities", JSONObject())
    PteIMCommerceCapability(data.optInt("version", 1), data.optString("currency", "COIN"), data.optBoolean("gifts"), data.optBoolean("backpack"), data.optBoolean("orders"), data.optBoolean("redPackets"))
  }

  fun gifts(callback: (Result<List<PteIMGift>>) -> Unit) = request(callback) {
    jsonList(sdk.postCommerceJson("/v1/commerce/gifts", JSONObject())) { PteIMGift.from(it) }
  }

  fun wallet(callback: (Result<PteIMWallet>) -> Unit) = request(callback) {
    PteIMWallet.from(sdk.postCommerceJson("/v1/commerce/wallet", JSONObject()))
  }

  fun backpack(callback: (Result<List<PteIMBackpackItem>>) -> Unit) = request(callback) {
    jsonList(sdk.postCommerceJson("/v1/commerce/backpack", JSONObject())) { PteIMBackpackItem.from(it) }
  }

  fun orders(limit: Int = 50, callback: (Result<List<PteIMCommerceOrder>>) -> Unit) = request(callback) {
    jsonList(sdk.postCommerceJson("/v1/commerce/orders", JSONObject().put("limit", limit))) { PteIMCommerceOrder.from(it) }
  }

  fun sendGift(request: PteIMGiftSendRequest, callback: (Result<PteIMCommerceOrder>) -> Unit) = request(callback) {
    PteIMCommerceOrder.from(sdk.postCommerceJson("/v1/commerce/gifts/send", request.toJson()))
  }

  fun useBackpack(request: PteIMBackpackUseRequest, callback: (Result<PteIMCommerceOrder>) -> Unit) = request(callback) {
    PteIMCommerceOrder.from(sdk.postCommerceJson("/v1/commerce/backpack/use", request.toJson()))
  }

  fun createRedPacket(request: PteIMRedPacketCreateRequest, callback: (Result<PteIMRedPacket>) -> Unit) = request(callback) {
    PteIMRedPacket.from(sdk.postCommerceJson("/v1/commerce/red-packets/create", request.toJson()))
  }

  fun redPacket(redPacketId: String, callback: (Result<PteIMRedPacketDetail>) -> Unit) = request(callback) {
    PteIMRedPacketDetail.from(sdk.postCommerceJson("/v1/commerce/red-packets/$redPacketId", JSONObject()))
  }

  fun claimRedPacket(redPacketId: String, callback: (Result<PteIMRedPacketDetail>) -> Unit) = request(callback) {
    PteIMRedPacketDetail.from(sdk.postCommerceJson("/v1/commerce/red-packets/$redPacketId/claim", JSONObject()))
  }

  private fun <T> request(callback: (Result<T>) -> Unit, action: () -> T) = sdk.executeCommerce { callback(runCatching(action)) }
  private fun <T> jsonList(value: JSONObject, mapper: (JSONObject) -> T): List<T> {
    val array = value.optJSONArray("list") ?: value.optJSONArray("data") ?: JSONArray()
    return (0 until array.length()).map { mapper(array.getJSONObject(it)) }
  }
}

data class PteIMCommerceCapability(val version: Int, val currency: String, val gifts: Boolean, val backpack: Boolean, val orders: Boolean, val redPackets: Boolean)
data class PteIMGift(val sku: String, val title: String, val coverUrl: String, val unitAmount: Long, val currency: String) { companion object { fun from(v: JSONObject) = PteIMGift(v.optString("sku"), v.optString("title"), v.optString("cover_url"), v.optLong("unit_amount"), v.optString("currency", "COIN")) } }
data class PteIMWallet(val balance: Long, val currency: String) { companion object { fun from(v: JSONObject) = PteIMWallet(v.optLong("balance"), v.optString("currency", "COIN")) } }
data class PteIMBackpackItem(val sku: String, val title: String, val coverUrl: String, val quantity: Long, val expiresAt: Long) { companion object { fun from(v: JSONObject) = PteIMBackpackItem(v.optString("sku"), v.optString("title"), v.optString("cover_url"), v.optLong("quantity"), v.optLong("expires_at")) } }
data class PteIMCommerceOrder(val orderId: String, val type: String, val resourceId: String, val amount: Long, val currency: String, val status: Int, val snapshot: String) { companion object { fun from(v: JSONObject) = PteIMCommerceOrder(v.optString("order_id"), v.optString("type"), v.optString("resource_id"), v.optLong("amount"), v.optString("currency", "COIN"), v.optInt("status"), v.optString("snapshot")) } }
data class PteIMRedPacket(val redPacketId: String, val roomId: String, val mode: String, val greeting: String, val totalAmount: Long, val totalCount: Int, val remainingAmount: Long, val remainingCount: Int, val currency: String, val status: Int, val expiresAt: Long) { companion object { fun from(v: JSONObject) = PteIMRedPacket(v.optString("red_packet_id"), v.optString("room_id"), v.optString("mode"), v.optString("greeting"), v.optLong("total_amount"), v.optInt("total_count"), v.optLong("remaining_amount"), v.optInt("remaining_count"), v.optString("currency", "COIN"), v.optInt("status"), v.optLong("expires_at")) } }
data class PteIMRedPacketClaim(val amount: Long) { companion object { fun from(v: JSONObject) = PteIMRedPacketClaim(v.optLong("amount")) } }
data class PteIMRedPacketDetail(val redPacket: PteIMRedPacket, val myClaim: PteIMRedPacketClaim?) { companion object { fun from(v: JSONObject): PteIMRedPacketDetail { val claim = if (v.has("myClaim") && !v.isNull("myClaim")) PteIMRedPacketClaim.from(v.getJSONObject("myClaim")) else if (v.has("claim")) PteIMRedPacketClaim.from(v.getJSONObject("claim")) else null; return PteIMRedPacketDetail(PteIMRedPacket.from(v.getJSONObject("redPacket")), claim) } } }
data class PteIMGiftSendRequest(val clientRequestId: String, val sku: String, val quantity: Long, val targetUserId: Long = 0, val sceneType: String = "", val roomId: String = "") { fun toJson() = JSONObject().put("clientRequestId", clientRequestId).put("sku", sku).put("quantity", quantity).put("targetUserId", targetUserId).put("sceneType", sceneType).put("roomId", roomId) }
data class PteIMBackpackUseRequest(val clientRequestId: String, val sku: String, val quantity: Long, val sceneType: String = "", val roomId: String = "") { fun toJson() = JSONObject().put("clientRequestId", clientRequestId).put("sku", sku).put("quantity", quantity).put("sceneType", sceneType).put("roomId", roomId) }
data class PteIMRedPacketCreateRequest(val clientRequestId: String, val roomId: String, val totalAmount: Long, val totalCount: Int, val mode: String = "lucky", val greeting: String = "", val sceneType: String = "live", val expiresIn: Long = 86400) { fun toJson() = JSONObject().put("clientRequestId", clientRequestId).put("roomId", roomId).put("totalAmount", totalAmount).put("totalCount", totalCount).put("mode", mode).put("greeting", greeting).put("sceneType", sceneType).put("expiresIn", expiresIn) }

File: ios/PteIMSDK/Sources/PteIMSDK/PteIMCommerce.swift
import Foundation

/** First-party Commerce extension. It reuses UserSig; it never creates a second IM connection. */
public final class PteIMCommerce: @unchecked Sendable {
  private unowned let sdk: PteIMSDK
  internal init(sdk: PteIMSDK) { self.sdk = sdk }

  public func capabilities() async throws -> PteIMCommerceCapability {
    try decode(try await sdk.commercePayload(path: "v1/commerce/capabilities", body: [:]))
  }
  public func gifts() async throws -> [PteIMGift] { try list(try await sdk.commercePayload(path: "v1/commerce/gifts", body: [:])) }
  public func wallet() async throws -> PteIMWallet { try decode(try await sdk.commercePayload(path: "v1/commerce/wallet", body: [:])) }
  public func backpack() async throws -> [PteIMBackpackItem] { try list(try await sdk.commercePayload(path: "v1/commerce/backpack", body: [:])) }
  public func orders(limit: Int = 50) async throws -> [PteIMCommerceOrder] { try list(try await sdk.commercePayload(path: "v1/commerce/orders", body: ["limit": limit])) }
  public func sendGift(_ request: PteIMGiftSendRequest) async throws -> PteIMCommerceOrder { try decode(try await sdk.commercePayload(path: "v1/commerce/gifts/send", body: request.json)) }
  public func useBackpack(_ request: PteIMBackpackUseRequest) async throws -> PteIMCommerceOrder { try decode(try await sdk.commercePayload(path: "v1/commerce/backpack/use", body: request.json)) }
  public func createRedPacket(_ request: PteIMRedPacketCreateRequest) async throws -> PteIMRedPacket { try decode(try await sdk.commercePayload(path: "v1/commerce/red-packets/create", body: request.json)) }
  public func redPacket(_ id: String) async throws -> PteIMRedPacketDetail { try decode(try await sdk.commercePayload(path: "v1/commerce/red-packets/\(id)", body: [:])) }
  public func claimRedPacket(_ id: String) async throws -> PteIMRedPacketDetail { try decode(try await sdk.commercePayload(path: "v1/commerce/red-packets/\(id)/claim", body: [:])) }

  private func decode<T: Decodable>(_ value: Any) throws -> T { try JSONDecoder().decode(T.self, from: JSONSerialization.data(withJSONObject: value)) }
  private func list<T: Decodable>(_ value: Any) throws -> [T] {
    guard let object = value as? [String: Any], let rows = object["list"] else { throw PteIMError.invalidResponse }
    return try JSONDecoder().decode([T].self, from: JSONSerialization.data(withJSONObject: rows))
  }
}

public struct PteIMCommerceCapability: Codable, Sendable { public let version: Int; public let currency: String; public let gifts: Bool; public let backpack: Bool; public let orders: Bool; public let redPackets: Bool }
public struct PteIMGift: Codable, Sendable { public let sku: String; public let title: String; public let coverURL: String; public let unitAmount: Int64; public let currency: String; enum CodingKeys: String, CodingKey { case sku, title, currency; case coverURL = "cover_url"; case unitAmount = "unit_amount" } }
public struct PteIMWallet: Codable, Sendable { public let balance: Int64; public let currency: String }
public struct PteIMBackpackItem: Codable, Sendable { public let sku: String; public let title: String; public let coverURL: String; public let quantity: Int64; public let expiresAt: Int64; enum CodingKeys: String, CodingKey { case sku, title, quantity; case coverURL = "cover_url"; case expiresAt = "expires_at" } }
public struct PteIMCommerceOrder: Codable, Sendable { public let orderID: String; public let type: String; public let resourceID: String; public let amount: Int64; public let currency: String; public let status: Int; public let snapshot: String; enum CodingKeys: String, CodingKey { case type, amount, currency, status, snapshot; case orderID = "order_id"; case resourceID = "resource_id" } }
public struct PteIMRedPacket: Codable, Sendable { public let redPacketID: String; public let roomID: String; public let mode: String; public let greeting: String; public let totalAmount: Int64; public let totalCount: Int; public let remainingAmount: Int64; public let remainingCount: Int; public let currency: String; public let status: Int; public let expiresAt: Int64; enum CodingKeys: String, CodingKey { case mode, greeting, currency, status; case redPacketID = "red_packet_id"; case roomID = "room_id"; case totalAmount = "total_amount"; case totalCount = "total_count"; case remainingAmount = "remaining_amount"; case remainingCount = "remaining_count"; case expiresAt = "expires_at" } }
public struct PteIMRedPacketClaim: Codable, Sendable { public let amount: Int64 }
public struct PteIMRedPacketDetail: Codable, Sendable { public let redPacket: PteIMRedPacket; public let myClaim: PteIMRedPacketClaim?; public let claim: PteIMRedPacketClaim? }
public struct PteIMGiftSendRequest: Sendable { public let clientRequestId: String; public let sku: String; public let quantity: Int64; public let targetUserId: Int64; public let sceneType: String; public let roomId: String; public init(clientRequestId: String, sku: String, quantity: Int64, targetUserId: Int64 = 0, sceneType: String = "", roomId: String = "") { self.clientRequestId = clientRequestId; self.sku = sku; self.quantity = quantity; self.targetUserId = targetUserId; self.sceneType = sceneType; self.roomId = roomId }; fileprivate var json: [String: Any] { ["clientRequestId": clientRequestId, "sku": sku, "quantity": quantity, "targetUserId": targetUserId, "sceneType": sceneType, "roomId": roomId] } }
public struct PteIMBackpackUseRequest: Sendable { public let clientRequestId: String; public let sku: String; public let quantity: Int64; public let sceneType: String; public let roomId: String; public init(clientRequestId: String, sku: String, quantity: Int64, sceneType: String = "", roomId: String = "") { self.clientRequestId = clientRequestId; self.sku = sku; self.quantity = quantity; self.sceneType = sceneType; self.roomId = roomId }; fileprivate var json: [String: Any] { ["clientRequestId": clientRequestId, "sku": sku, "quantity": quantity, "sceneType": sceneType, "roomId": roomId] } }
public struct PteIMRedPacketCreateRequest: Sendable { public let clientRequestId: String; public let roomId: String; public let totalAmount: Int64; public let totalCount: Int; public let mode: String; public let greeting: String; public let sceneType: String; public let expiresIn: Int64; public init(clientRequestId: String, roomId: String, totalAmount: Int64, totalCount: Int, mode: String = "lucky", greeting: String = "", sceneType: String = "live", expiresIn: Int64 = 86400) { self.clientRequestId = clientRequestId; self.roomId = roomId; self.totalAmount = totalAmount; self.totalCount = totalCount; self.mode = mode; self.greeting = greeting; self.sceneType = sceneType; self.expiresIn = expiresIn }; fileprivate var json: [String: Any] { ["clientRequestId": clientRequestId, "roomId": roomId, "totalAmount": totalAmount, "totalCount": totalCount, "mode": mode, "greeting": greeting, "sceneType": sceneType, "expiresIn": expiresIn] } }
